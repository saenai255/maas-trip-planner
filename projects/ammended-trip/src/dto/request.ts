import { APIGatewayEvent } from 'aws-lambda';
import { Coordinates } from './coordinates';
import BadRequestError from '../errors/bad-request';

type Request = {
    from: Coordinates;
    to: Coordinates;
};

function parseRequest(event: APIGatewayEvent): Request {
    const from = event.queryStringParameters?.from;
    const to = event.queryStringParameters?.to;

    if (!from || !to) {
        throw new BadRequestError('Missing required query parameters.', {
            from,
            to,
        });
    }

    const fromCoordinates = from.split(',').map((coordinate) => parseFloat(coordinate));
    const toCoordinates = to.split(',').map((coordinate) => parseFloat(coordinate));

    const coordsAreValidNumbers = [...fromCoordinates, ...toCoordinates].every(
        (coordinate) => !isNaN(coordinate) && typeof coordinate === 'number',
    );
    if (fromCoordinates.length !== 2 || toCoordinates.length !== 2 || !coordsAreValidNumbers) {
        throw new BadRequestError('Invalid query parameters.', {
            from,
            to,
        });
    }

    return {
        from: fromCoordinates as Coordinates,
        to: toCoordinates as Coordinates,
    };
}

export { parseRequest };

export default Request;
