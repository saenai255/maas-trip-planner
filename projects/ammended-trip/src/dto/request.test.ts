import { APIGatewayEvent } from 'aws-lambda';
import { parseRequest } from './request';

describe('Request', () => {
    describe('parseRequest', () => {
        it('should parse the request', () => {
            const event: APIGatewayEvent = {
                queryStringParameters: {
                    from: '60.1699,24.9384',
                    to: '60.1699,24.9384',
                },
            } as any;

            const result = parseRequest(event);
            expect(result).toEqual({
                from: [60.1699, 24.9384],
                to: [60.1699, 24.9384],
            });
        });

        it('should throw when given too few coordinates', () => {
            const event: APIGatewayEvent = {
                queryStringParameters: {
                    from: '60.1699',
                    to: '60.1699,24.9384',
                },
            } as any;

            expect(() => parseRequest(event)).toThrowErrorMatchingSnapshot();
        });

        it('should throw when given too many coordinates', () => {
            const event: APIGatewayEvent = {
                queryStringParameters: {
                    from: '60.1699,123,321',
                    to: '60.1699,24.9384',
                },
            } as any;

            expect(() => parseRequest(event)).toThrowErrorMatchingSnapshot();
        });

        it('should throw when given invalid coordinates', () => {
            const event: APIGatewayEvent = {
                queryStringParameters: {
                    from: 'X,24.9384',
                    to: '60.1699,24.9384',
                },
            } as any;

            expect(() => parseRequest(event)).toThrowErrorMatchingSnapshot();
        });
    });
});
