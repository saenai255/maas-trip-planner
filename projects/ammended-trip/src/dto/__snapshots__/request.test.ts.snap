// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Request parseRequest should throw when given invalid coordinates 1`] = `"Invalid query parameters."`;

exports[`Request parseRequest should throw when given too few coordinates 1`] = `"Invalid query parameters."`;

exports[`Request parseRequest should throw when given too many coordinates 1`] = `"Invalid query parameters."`;
